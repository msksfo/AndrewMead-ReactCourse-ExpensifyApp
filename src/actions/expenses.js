//* pseodo code for how my action generators initially worked
// 1. component calls the action generator
// 2. action generator returns an object (one of the objects below, or one         of the objects in the filters file)
// 3. the component takes the that object and dispatches it
// 4. the redux store runs the reducers and it changes

//* to use firebase, we will tweak how the action generators work
// it is a little more complex because we are working with asynchronous code
// 1. component calls the action generator
// 2. action generator returns a function
// 3. the component dispatches the function(?) //! redux by default does not allow you to dispatch function. to do that we have to set up a module, a piece of redux middleware, that is going to add support for this behavior
//* redux-thunk is the module mentioned above that provides support for dispatching functions
// 4. the function runs (it has the ability to dispatch other actions and do       whatever it wants) //* this is where we will put our firebase code

//* broad overview of working with firebase & redux
// (importing firebase, then: )
// 1. use push to add something to the database
// 2. then we will dispatch an action that returns an object, and THAT will manipulate the redux store

//notes related to the project as a whole
//* initial structure example (this code is not being used!)
/*
const db = {
    expenses: {
        asdfj: {},
    },
};
*/

//* goal structure of project, example (this code is not being used!)
// when reading/writing from expenses, it will be users/userID/expenses
//! we changed this over  in lecture 168
/*
const finalDB = {
    users: {
        theUserId: {
            expenses: {
                asdfj: {},
            },
        },
    },
};
*/

import uuid from 'uuid';
import database from '../firebase/firebase';

//set up the action generator for adding an expense
// destructure the values that come from the user with default values in case        none are passed in
const addExpense = expense => ({
    type: 'ADD_EXPENSE',
    expense,
});

// this will dispatch addExpense inside the function and that is what will keep changing the store
const startAddExpense = (expenseData = {}) => {
    // this returns the thing that gets dispatched. (before, we were just returning objects)
    //* this returned function gets called internally by redux, and it is called with dispatch
    //! lecture 168 update: to allow multiple users, pass getState in, to get the userid
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        // write data to firebase
        // wait for the data to correctly sync up
        // then we use dispatch to dispatch addExpense, making sure the redux store reflects those changes
        const {
            description = '',
            noteText = '',
            amount = 0,
            createdAt = 0,
        } = expenseData;

        const expense = { description, noteText, amount, createdAt };

        // access firebase and use .push() to save the data
        // in lecture 153, we altered this to *return* databse... By returning the promise chain, we can add .then() in the test file
        return (
            database
                //! what we pass into .ref() changed in lecture 168. it used to be .ref('expenses') but we want to have multiple users each with their own expenses
                .ref(`users/${uid}/expenses`)
                .push(expense)
                .then(ref => {
                    //! dispatch the action from up above
                    // otherwise the redux store will never be updated
                    dispatch(
                        addExpense({
                            id: ref.key,
                            ...expense,
                        })
                    );
                })
        );
    };
};

// set up the action generator for removing an expense by id
// also could pass in the id as a destructured value { id }
const removeExpense = id => ({
    type: 'REMOVE_EXPENSE',
    id,
});

const startRemoveExpense = id => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .remove()
            .then(() => {
                //* dispatch remove expenses
                dispatch(removeExpense(id));
            });
    };
};

// set up the action generator for editing an expense
// pass in an id, and an updates object
const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates,
});

const startEditExpense = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database
            .ref(`users/${uid}/expenses/${id}`)
            .update(updates)
            .then(() => {
                //* dispatch edit expenses
                dispatch(editExpense(id, updates));
            });
    };
};

// this will manipulate the redux store in some way
//* get the array back from firebase, and set it
const setExpenses = expenses => ({
    type: 'SET_EXPENSES',
    expenses,
});

// this is the asynchronous action. it fetches the data
const startSetExpenses = () => {
    //1. fetch all expense data once
    //2. parse the data into an array
    //3. dispatch set_expenses

    //* this first returned function is a requirement of redux-thunk. redux-thunk requires that you dispatch a function, so i need to return a function from the action generator. that funnction is then called and this is when we start manipulating the firebase data
    return (dispatch, getState) => {
        const uid = getState().auth.uid;

        //* fetch all expense data once
        return database
            .ref(`users/${uid}/expenses`)
            .once('value')
            .then(snapshot => {
                //console.log(snapshot);

                const expenses = [];

                //* parse the data (currently an object) into an array
                // this function will be called one time for each expense
                snapshot.forEach(childSnapshot => {
                    //console.log(childSnapshot.val());

                    // push the child object into the expenses array
                    //* then we have our array of objects that we can use as redux expects us to in this application
                    expenses.push({
                        // grab the id generated by firebase with .key
                        id: childSnapshot.key,
                        ...childSnapshot.val(),
                    });
                });
                //console.log(expenses);
                //* dispatch set expenses
                dispatch(setExpenses(expenses));
            });
    };
};

export {
    setExpenses,
    startSetExpenses,
    addExpense,
    removeExpense,
    editExpense,
    startAddExpense,
    startRemoveExpense,
    startEditExpense,
};
