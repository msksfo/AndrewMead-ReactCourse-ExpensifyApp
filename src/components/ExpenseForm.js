import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

//console.log(now.format('MMM Do, YYYY'));

// this contains all the form fields and logic to get the form to work
export default class ExpenseForm extends React.Component {
    // use local component state to track the changes to these form inputs. only when the user submits the form will we send it off to redux. redux will figure out what to do (add a new expense, or edit the existing expense)

    //! WHY DO WE USE A CONSTRUCTOR FUNCTION HERE?
    //! we need to look at the props (for our ternary conditions). so we need to use a constructor function, and define the state there. then we can pass the props up to super
    constructor(props) {
        super(props);

        // *we only want to use defaults for the first 4 values if they are CREATING an expense. if they are EDITING an expense, we need to know what their value were so we can populate the form fields with that information.

        this.state = {
            description: props.expense ? props.expense.description : '',
            noteText: props.expense ? props.expense.noteText : '',
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : '',
            createdAt: props.expense
                ? moment(props.expense.createdAt)
                : moment(),
            calendarFocused: false,
            error: '',
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onAmountChange = e => {
        const amount = e.target.value;
        // ensure that the number comes in a specific format, with regex

        // myRegex makes sure the value starts with a number ^\d
        // force usage of at least one number ^\d{1,}
        // optionally allow for one (only one!) decimal point  ^\d*(\.)?
        // allow only 2 numbers after the decimal point ^\d*(\.\d{0,2})?
        // make sure the regex ends with 2 numbers after the decimal point (if /when present) ^\d*(\.\d{0,2})?$

        const myRegex = /^\d{1,}(\.\d{0,2})?$/;

        // if there is no amount, OR the amount provided is a match
        if (!amount || amount.match(myRegex)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        // call this function if a date was picked
        // if it was cleared, it will be called with nothing
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };

    //?  I dont really understand this function. Review Andrew's explanation in section 11, lecture 106 or read docs
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };

    onNoteChange = e => {
        //! trying to access e.target.value inside the callback to setState will not work!

        //* always make a variable for this (e.target.value)
        // the other option would be to use e.persist()
        const noteText = e.target.value;
        this.setState(() => ({ noteText }));
    };

    onSubmit = e => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            // set error state equal to 'please provide amount and description'
            this.setState(() => ({
                error: 'please provide amount and description',
            }));
        } else {
            // clear the error
            this.setState(() => ({ error: '' }));

            // submit the form
            this.props.onSubmit({
                description: this.state.description,
                // turn the string into a decimal number, and multiply by 100 to get dollars from our cents
                amount: parseFloat(this.state.amount, 10) * 100,
                // parse this to get the timestamp
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note,
            });
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        // adding value makes it a read only input. to make it dynamic as the user types, add the onChange handler
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                        placeholder="Description"
                        autoFocus
                    />

                    <input
                        type="text"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                        placeholder="Amount"
                    />

                    <SingleDatePicker
                        //*the first 4 properties are mandatory
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        // make all days available, even dates in the past
                        isOutsideRange={() => false}
                    />

                    <textarea
                        placeholder="Add a note for your stupid fucking expense (fucking optional)"
                        value={this.state.noteText}
                        onChange={this.onNoteChange}
                    />

                    <button>Add Expense</button>
                </form>
            </div>
        );
    }
}
