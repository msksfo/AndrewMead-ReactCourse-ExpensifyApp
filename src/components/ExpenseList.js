import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//*  use the connect function in the individual components where you need to dispatch actions or read from the store. this connects your component to the redux store

//* when you connect a component to the redux store, it's reactive. which means that as the store changes, the component gets rerendered with the new values

// this is being exported so that we can test it
export const ExpenseList = props => (
    <div>
        {props.expenses.length === 0 ? (
            <p>no expenses</p>
        ) : (
            props.expenses.map(value => (
                <ExpenseListItem
                    key={value.id}
                    id={value.id}
                    description={value.description}
                    amount={value.amount}
                    createdAt={value.createdAt}
                />
            ))
        )}
    </div>
);

//! STEPS TO CONNECT A COMPONENT TO THE REDUX STORE
// a) import { connect } from 'react-redux'
// 1. create a const for the higher order component by calling connect()
// 2. connect() returns a function, not a higher order component
// 3. to get the HOC, we need to call the function that connect returns, passing        in our component ->
// 4. connect()(ourComponent)
// 5. pass a function into connect that says what information from the store we         want our component to be able to access connect(()=> {})(ourComponent)
// 6. the argument to the anonymous function above is state. the information            returned is an object of key/value pairs usually from state, but we can           also add our own -> connect((state) => {name: 'harry'})(ourComponent)

//* export default connect()() is more common practice than const blahblah = connect()()
export default connect(state => {
    // what information from the store do we want our component to be able to access. the information will be returned as an object of key/value pairs

    return {
        // add a new property, like name
        name: 'cirrus',

        // or use the ones from state, like expenses and filters
        //expenses: state.expenses,
        //filters: state.filters,

        //* refactored to use the return value (the filtered sorted array) of getVisibleExpenses function, which was a default export, so here we imported it as selectExpenses

        expenses: selectExpenses(state.expenses, state.filters),
    };
})(ExpenseList);

//export default ConnectedExpenseList;

//* it is common practice to extract the function passed to connect() to it's own variable, const mapStateToProps, as follows:
/* 
    -named this way because it maps the store state to component props

    const mapStateToProps = (state) => {
        return {
            name: 'cirrus',
            expenses: state.expenses,
        };
    }

    - then connect becomes connect(mapStateToProps)(ourComponent)
    ! in summary, call connect, define the things we want to get off of the store, define the component we want to create the connected version of
    * the end result is a new component with the props from the store
*/
