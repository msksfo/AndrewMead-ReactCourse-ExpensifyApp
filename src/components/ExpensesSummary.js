import React from 'react';
import getVisibleExpenses from '../selectors/expenses';
import getTotalExpenses from '../selectors/expenses-total';
import { connect } from 'react-redux';
import numeral from 'numeral';

//* could have destructured off of the props object like this:
// const ExpenseSummary = ({ expenseCount, expenseTotal}) => {}
export const ExpensesSummary = props => {
    const expenseWord = props.expenses.length === 1 ? 'expense' : 'expenses';
    const formattedExpenseTotal = numeral(
        getTotalExpenses(props.expenses) / 100
    ).format('$0,0.00');

    return (
        <div>
            <h3>
                Viewing {props.expenses.length} {expenseWord} totalling{' '}
                {formattedExpenseTotal}
            </h3>
        </div>
    );
};

const mapStateToProps = state => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters),
    };

    //* could also have done it this way:
    /*
        const visibleExpenses = getVisibleExpenses(state.expenses, state.filters

        return {
            expenseCount: visibleExpenses.length,
            expensesTotal: getTotalExpenses(visibleExpenses)
        }
    */
};

export default connect(mapStateToProps)(ExpensesSummary);
