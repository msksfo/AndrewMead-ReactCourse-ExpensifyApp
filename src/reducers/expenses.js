const expensesReducerDefaultState = [];

// pass in the default state, an empty array. and pass in the action
const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // add the expense to the array of expenses
            // return state.concat(action.expense) -> with es6 array spread ->
            return [...state, action.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(value => value.id !== action.id);

        //* the above return statement destructured ->
        // return state.filter({ id } => id !== action.id)

        case 'EDIT_EXPENSE':
            return state.map(value => {
                return value.id === action.id
                    ? { ...value, ...action.updates }
                    : value;
            });

        case 'SET_EXPENSES':
            // we don't care what is in state, or what the previous expenses were
            // we just want to set the expenses array completely
            return action.expenses;

        default:
            return state;
    }
};

export default expensesReducer;
