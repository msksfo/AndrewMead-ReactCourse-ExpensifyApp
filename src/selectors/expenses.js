import moment from 'moment';

const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
    // get a subset of all the expenses
    return expenses
        .filter(value => {
            const createdAtMoment = moment(value.createdAt);

            const startDateMatch = startDate
                ? startDate.isSameOrBefore(createdAtMoment, 'day')
                : true;

            const endDateMatch = endDate
                ? endDate.isSameOrAfter(createdAtMoment, 'day')
                : true;

            // filter for expenses.description has the text variable string inside it
            // hint: use .includes() and convert both strings to lower case
            const textMatch = value.description
                .toLowerCase()
                .includes(text.toLowerCase());

            // add the expense to the filtered array if ALL three of these are true
            return startDateMatch && endDateMatch && textMatch;
        })
        .sort((a, b) => {
            if (sortBy === 'date') {
                // the bigger number is the more recently created
                return a.createdAt > b.createdAt ? -1 : 1;
            } else if (sortBy === 'amount') {
                // put most expensive expenses first
                return a.amount > b.amount ? -1 : 1;
            }
        });
};

export default getVisibleExpenses;
