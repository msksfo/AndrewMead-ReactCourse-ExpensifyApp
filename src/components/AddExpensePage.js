import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// export this (unconnected) so we can test it
export class AddExpensePage extends React.Component {
    onSubmit = expense => {
        // dispatch the action to the redux store by calling connect
        //props.dispatch(addExpense(expense));

        //* the above code was refactored to this after adding in mapDispatchToProps, which we did for testing purposes
        this.props.startAddExpense(expense);

        // force a redirect to the dashboard page when user submits
        // pass in the path to the page you want to go to
        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        );
    }
}

const mapDispatchToProps = dispatch => {
    // return an object with props. these props call dispatch
    return {
        startAddExpense: expense => dispatch(startAddExpense(expense)),
    };
};

export default connect(
    undefined,
    mapDispatchToProps
)(AddExpensePage);

/* 
     this was the original code: export default connect()(AddExpensePage);

    -because addExpense is a function we imported in, it is difficult to test (unlike if it were passed in as a prop). but connect gives us a work around. 

    so we will refactor the export to
    1.  export default connect(undefined, mapDispatchToProps)(AddExpensePage)

    2. note: we also changed the code inside the arrow function passed to onSubmit from this:  //props.dispatch(addExpense(expense))
                to this:  //props.onSubmit(expense)

    3. we changed it over to a class based component, to avoid inline functions

    the functionality is the same! it was only changed so we could test with jest



*/
