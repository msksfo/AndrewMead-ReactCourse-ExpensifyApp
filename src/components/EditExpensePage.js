import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';

// import my action generators for removing/editing expenses
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = expense => {
        this.props.startEditExpense(this.props.expense.id, expense);

        // history.push() navigates the user programatically, after some code has run. Link is an anchor tag, so nothing happens unless it is clicked
        this.props.history.push('/');
    };

    onRemove = () => {
        this.props.startRemoveExpense(this.props.expense.id);

        this.props.history.push('/');
    };

    render() {
        return (
            <div>
                <ExpenseForm
                    // pass the expense object down (it was returned from  mapStateToProps), so that when the create expense page comes up, the field will be populated with the data that previously existed, so the user can edit it
                    expense={this.props.expense}
                    // dispatch the action to edit the expense
                    // redirect to the dashboard page
                    onSubmit={this.onSubmit}
                />

                <button
                    onClick={this.onRemove}
                    // we have access to dispatch (as a prop) inside our connected components
                    // use dispatch to update the store as users interact with our app
                    // pass in the action object
                    // remove an expense by dispatching an action when the button is clicke
                >
                    Remove
                </button>
            </div>
        );
    }
}

//* 1. react router (the Link that led us here) renders the higher order component
//* 2. the hoc passes the props through and allows us to add some new ones

const mapStateToProps = (state, props) => {
    //* render one new prop
    // use the passed in props to search through the expenses array
    return {
        expense: state.expenses.find(
            value => value.id === props.match.params.id
        ),
    };

    //! now the actual expense we are trying to edit comes back to us, so up above we can render an instance of Expense form, passing in this specific expense as a prop
};

//? why is props being passed in? i dont see that it's being used...
const mapDispatchToProps = (dispatch, props) => {
    return {
        // take the data that comes in, -----> pass it down
        startEditExpense: (id, expense) =>
            dispatch(startEditExpense(id, expense)),
        startRemoveExpense: data => dispatch(startRemoveExpense(data)),
    };
};

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(EditExpensePage);
