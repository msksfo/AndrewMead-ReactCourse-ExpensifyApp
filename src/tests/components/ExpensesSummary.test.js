import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';

test('should correctly render expenses summary with one expense', () => {
    const singleExpense = [expenses[0]];
    const wrapper = shallow(<ExpensesSummary expenses={singleExpense} />);
    expect(wrapper).toMatchSnapshot();
});

test('should correctly render expenses summary with multiple expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});
