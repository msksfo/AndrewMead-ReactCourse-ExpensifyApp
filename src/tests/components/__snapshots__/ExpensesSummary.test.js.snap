// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render expenses summary with multiple expenses 1`] = `
<div>
  <h3>
    Viewing 
    3
     
    expenses
     totalling
     
    $1,141.95
  </h3>
</div>
`;

exports[`should correctly render expenses summary with one expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling
     
    $1.95
  </h3>
</div>
`;
